Add `signed_by_[any|all]` parameters to `gpg.verify`

lkubb opened the pull request:
### What does this PR do?
Allows `gpg.verify` to additionally verify that good signatures come from a set of predefined public keys.

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/63166

### Previous Behavior
Any key present in the keyring can provide a valid signature for anything.

### New Behavior
It is possible to restrict the accepted signers to a predefined set of keys.

### Notes
This is based off of https://github.com/saltstack/salt/pull/63152, so should be merged after that one.

### Merge requirements satisfied?
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

lkubb committed:
Add test for issue 63145

lkubb committed:
Make gpg.verify respect gnupghome

lkubb committed:
Add `signed_by_any`/`signed_by_all` params to gpg.verify

lkubb committed:
Reconsider status check

lkubb committed:
Fix tests

lkubb committed:
Plug verification issue after status check reconsideration

Since missing pubkeys also cause the previous signature's
fingerprint to be overwritten, this would have led to
situations where the check could have passed when it
should not have.

lkubb committed:
Cleanup logic

lkubb committed:
Workaround python-gnupg issue

https://github.com/vsajip/python-gnupg/issues/214

lkubb committed:
Adapt offical fix for workaround

https://github.com/vsajip/python-gnupg/commit/ee94a7ecc1a86484c9f02337e2bbdd05fd32b383
That was very fast.

lkubb committed:
Backport more meaningful test

lkubb committed:
Merge branch 'gpg-verify-gnupghome' into gpg-signed-by

lkubb committed:
Update versionadded

lkubb committed:
Avoid exception on import w/ missing gnupg lib

Ch3LL commented on salt/modules/gpg.py:
```diff
@@ -1393,3 +1502,22 @@ def decrypt(
+
+    class FixedVerify(gnupg.Verify):
+        """
+        This is a workaround for https://github.com/vsajip/python-gnupg/issues/214.
```
Looks like you have pointed to an issue that has been closed and fixed. Do you still need this for newer versions of python-gnupg?

lkubb replied:
It's unnecessary for upcoming versions, but does not interfere with them either. Would you prefer this workaround not be applied to those? (Currently, all released versions that support `sig_info` have this bug since I discovered it working on this feature.)

Ch3LL reviewed (commented):
Just a question

lkubb committed:
Merge branch 'master' into gpg-signed-by

lkubb committed:
Do not apply workaround on fixed versions

lkubb commented:
Huh, I accidentally performed a squash merge from master, which blew something up on Github. In the meantime, I force-pushed a normal merge, but somehow the first one triggered another reviewer to be requested as well. Sorry! =/

The new commit only applies the workaround to versions `0.5.0` and below (all currently released ones) and sets `versionadded` back to 3006.

lkubb committed:
Correct bool comparison

garethgreenaway commented:
@lkubb Just a quick question, are the two options `signed_by_any` and `signed_by_all` mutually exclusive or are they able to be specified together in a function call?  I didn't see a check for the scenario when they're both included but wanted to double check.

lkubb commented:
@garethgreenaway They are able to be specified together:

https://github.com/lkubb/salt/blob/809c146c5a062cd9cf3efb07bb95f13c78d55435/salt/modules/gpg.py#L1286-L1289

garethgreenaway commented:
@lkubb if it's not a issue to include both I think it's fine.

lkubb committed:
Merge branch 'master' into gpg-signed-by

lkubb committed:
Account for subkeys

sig_info["fingerprint"] contains the actual signing key's fingerprint,
which might be a subkey. The primary key's fingerprint is always found in
sig_info["pubkey_fingerprint"]. In cases where a signing subkey was
used, the intended behavior is still comparison with the primary key.

lkubb committed:
Bump versionadded

lkubb commented:
@garethgreenaway Sorry for dismissing the reviews, I had to push two slight modifications:
* `versionadded` was `3006.0`, but this PR is tagged with `3007.0`
* The comparison was done with the actual signing key's fingerprint which – in cases where a signing subkey was used – differs from the primary key's fingerprint. The expected behavior is to compare to the primary key, I assume. This just came up during usage.

lkubb committed:
Merge branch 'master' into gpg-signed-by

lkubb committed:
Rename changelog files to .md

lkubb committed:
Merge branch 'master' into gpg-signed-by
